open-craft/blockstore
# Make command to run an isolated, ephemeral blockstore server for tests and development

## Description

This adds a new make command, `make testserver` which you can run on your host to spin up a temporary, isolated instance of blockstore that runs at http://localhost:18251 . The idea is that this isolated instance can be used by edx-platform for automated tests of functionality that depends on Blockstore.

The tests should be written so that they do not assume any cleanup is done automatically; each test (class) should create bundles/collections when needed without ever using fixed UUIDs. That way, for example, a single Blockstore server can be used as a target for all test runs on Jenkins, a single instance of this `make testserver` command can be used for many separate local runs of the edxapp test suite during development, and tests can be parallelized.

Simply hit CTRL-C when the server is no longer useful (you're done running tests) and its container, MySQL database, and blockstore data (media) will be destroyed. Only the virtualenv volume is kept, to make subsequent runs of `make testserver` much faster.

## Author Comments, Concerns, and Open Questions

I have set the target to be `ormsbee/links` temporarily because I need to use this to write tests that may depend on the links functionality, but I don't want it to appear in the diffs.

I'm thinking that for now, any test we write in edxapp that depends on blockstore can be feature-flagged and off by default? Eventually we'll want them running all the time, but to do that we need a version of this container that's always running on devstack and jenkins (or at least is spun up while the test suite runs).

The Dockerfiles and docker-compose files here are _much_ faster and easier to work with than the ones in `edx/configuration` and `edx/devstack`, but I assume we'll have to transition to the latter - see https://github.com/edx/configuration/pull/5352 for a start

## Test Instructions

Optional: Run a normal instance of blockstore (`make dev.up`, `make blockstore-shell`, `make runserver`)

Run `make testserver` on the host computer.

You should now have an isolated instance of blockstore. For debugging convenience, you can log in using `staff` / `edx` at http://localhost:18251/admin/ . Confirm this instance is using a separate database, and that an auth token was created at http://localhost:18251/admin/authtoken/token/ for edx-platform to use.
